### ISSUE TITLE
Arbitrary Gen of a big case class

### ISSUE THREAD

**Issue by NicolasRiou** [typelevel/scalacheck#489]
Hello,

I am getting an error when I try to generate test data for my case class. I am using this : 

**val dataArb2 = Arbitrary(Gen.resultOf(PatrimoineDiff))**

in order to generate dummy data without too much effort. 

And my case case is defined like this: 

```
import org.scalatest.FunSuite
import org.scalacheck.{Arbitrary, Gen}
import java.util.Random

case class PatrimoineDiff (
  ReadingPointId: Int
  , DateDerniereMaj: java.sql.Timestamp
  , ShortName: String
  , Estate_Name: String
  , AddressLine1: String
  , AddressLine2: String
  , PostCode: String
  , Town: String
  , Latitude: String
  , Longitude: String
  , Country_Name: String
  , Entrance_OrderNumber: String
  , Building: String
  , Flat_OrderNumber: Int
  , FlatType_Description: String
  , Surface: Double
  , FlatNumber: String
  , Tantieme: Double
  , ReadingPoint_OrderNumber: Int
  , ReadingType_Description: String
  , FluidType: String
  , DeviceId: Int
  , SerialNumber: String
  , SerialNumber2: String
  , ForcedCorrectionRate: Double
  , Description: String
  , PulseValue: Float
  , AMRNumber: String
  , AMRNumber2: String
  , Make: String
  , Model: String
  , IndexPose: Float
  , Disabled: Int
)


implicit def arbDate: Arbitrary[java.sql.Timestamp] = {
    val minValue : Int = 0
    val maxValue : Int = 10000000
    val r = new Random();
    Arbitrary(new java.sql.Timestamp(r.nextInt((maxValue - minValue) + 1) + minValue))
}

val dataArb2 = Arbitrary(Gen.resultOf(PatrimoineDiff))
```

I am getting this error (this is just the end of the error message because it actually takes a whole screen) : 

```
..
[T1, T2, T3, T4, T5, R](f: (T1, T2, T3, T4, T5) => R)(implicit a1: org.scalacheck.Arbitrary[T1], implicit a2: org.scalacheck.Arbitrary[T2], implicit a3: org.scalacheck.Arbitrary[T3], implicit a4: org.scalacheck.Arbitrary[T4], implicit a5: org.scalacheck.Arbitrary[T5])org.scalacheck.Gen[R] <and>
  [T1, T2, T3, T4, R](f: (T1, T2, T3, T4) => R)(implicit a1: org.scalacheck.Arbitrary[T1], implicit a2: org.scalacheck.Arbitrary[T2], implicit a3: org.scalacheck.Arbitrary[T3], implicit a4: org.scalacheck.Arbitrary[T4])org.scalacheck.Gen[R] <and>
  [T1, T2, T3, R](f: (T1, T2, T3) => R)(implicit a1: org.scalacheck.Arbitrary[T1], implicit a2: org.scalacheck.Arbitrary[T2], implicit a3: org.scalacheck.Arbitrary[T3])org.scalacheck.Gen[R] <and>
  [T1, T2, R](f: (T1, T2) => R)(implicit a1: org.scalacheck.Arbitrary[T1], implicit a2: org.scalacheck.Arbitrary[T2])org.scalacheck.Gen[R]
 cannot be applied to (PatrimoineDiff.type)

```

Now the weird thing is that it works as intended if i take off a few properties from my case class :
```


case class PatrimoineDiff (
  ReadingPointId: Int
  , DateDerniereMaj: java.sql.Timestamp
  , ShortName: String
  , Estate_Name: String
  , AddressLine1: String
  , AddressLine2: String
  , PostCode: String
  , Town: String
  , Latitude: String
  , Longitude: String
  , Country_Name: String
  , Entrance_OrderNumber: String
  , Building: String
  , Flat_OrderNumber: Int
  , FlatType_Description: String
  , Surface: Double
  , FlatNumber: String
  , Tantieme: Double
  , ReadingPoint_OrderNumber: Int
  , ReadingType_Description: String
  , FluidType: String
  , DeviceId: Int
)


implicit def arbDate: Arbitrary[java.sql.Timestamp] = {
  val minValue : Int = 0
  val maxValue : Int = 10000000
  val r = new Random();
  Arbitrary(new java.sql.Timestamp(r.nextInt((maxValue - minValue) + 1) + minValue))
}

val dataArb2 = Arbitrary(Gen.resultOf(PatrimoineDiff))

```

This works fine but adding one more fields gives me the error back. So is there some kind of limitation somewhere on the number of fields we can have in a case class. 

Thanks for helping.

**Comment by ashawley** [comment]
This is hard to follow, because your post isn't formatted with GitHub-flavored Markdown.

It sounds like ScalaCheck only supports case classes of 22 fields.

**Comment by NicolasRiou** [comment]
I and thanks for your feedback. I put the code in bold for readability. 
I ended up generating each field one by one like this : 

```
var generator: PatrimoineDiff = PatrimoineDiff(
    Gen.choose(0, 300000).sample.get,
    arbDate.arbitrary.sample.get,
    Arbitrary.arbString.arbitrary.sample.get,
    ..... an so on .... 
    Arbitrary.arbString.arbitrary.sample.get,
    Arbitrary.arbString.arbitrary.sample.get,
    Arbitrary.arbFloat.arbitrary.sample.get,
    Arbitrary.arbInt.arbitrary.sample.get
  )
```

**Comment by ashawley** [comment]
That doesn't seem like the best way to do it, but you're on the right track.

The way ScalaCheck implemented `Gen.resultOf` won't support classes of more than 22 fields. 
 
You should investigate https://github.com/alexarchambault/scalacheck-shapeless/ which should be able to provide an `Arbitrary` for classes of more than 22 fields with minimal boilerplate.  I've never had to write property tests for large classes, so I'm not sure how well it works.   Since you'll probably need to narrow the range of values for the fields of your large class, you'll need to write your own boilerplate for `Gen` and `Arbitrary.  You could _spin the wheel_ with scalacheck-shapeless and see what happens.

There is a better way to write your own `Arbitrary` for a large class.  I won't do it for your particular class, but here is how to do it for a class of 23 `Int`s.

```scala
val genClassOf23: Gen[ClassOf23] =
  for {
    i1 <- Arbitrary.arbitrary[Int]
    i2 <- Arbitrary.arbitrary[Int]
    i3 <- Arbitrary.arbitrary[Int]
    ...
    i21 <- Arbitrary.arbitrary[Int]
    i22 <- Arbitrary.arbitrary[Int]
    i23 <- Arbitrary.arbitrary[Int]
  } yield {
    ClassOf23(
      i1,i2,i3,i4,i5,i6,i7,i8,i9,i10,i11,i12,i13,
      i14,i15,i16,i17,i18,i19,i20,i21,i22,i23
    )
  }

implicit val arbClassOf23: Arbitrary[ClassOf23] =
  Arbitrary(genClassOf23)

property("arbitrary[ClassOf23]") = Prop.forAll { _: ClassOf23 =>
  Prop.proved
}
```
